cmd/admin-heal-ops.go

```diff
@@ -0,0 +1,659 @@
+
+	// map of healed parts count to number of objects with that
+	// many healed parts
+	ObjectsByHealedPC map[int]int64 `json:"ObjectsByHealedPartCount"`
+
+	// map of disk serial number to number of objects with parts
+	// missing on that disk
+	ObjectsMissingByDisk map[int]int64 `json:"ObjectsMissingByDisk"`
+
+	// map of disk serial number to number of objects available on
+	// that disk
+	ObjectsAvailableByDisk map[int]int64 `json:"ObjectsAvailableByDisk"`
+
+	// map of disk serial number to number of objects healed on
+	// that disk
+	ObjectsHealedByDisk map[int]int64 `json:"ObjectsHealedByDisk"`
+}
+
+// newHealStatistics - creates heal stats object given number of disks
+func newHealStatistics(numDisks int) healStatistics {
+	return healStatistics{
+		NumDisks:               numDisks,
+		BucketsAvailableByDisk: make(map[int]int64, numDisks),
+		BucketsMissingByDisk:   make(map[int]int64, numDisks),
+		BucketsHealedByDisk:    make(map[int]int64, numDisks),
+		ObjectsByAvailablePC:   make(map[int]int64, numDisks),
+		ObjectsByHealedPC:      make(map[int]int64, numDisks),
+		ObjectsMissingByDisk:   make(map[int]int64, numDisks),
+		ObjectsAvailableByDisk: make(map[int]int64, numDisks),
+		ObjectsHealedByDisk:    make(map[int]int64, numDisks),
+	}
+}
+
+// makeIntSets - creates a int set from a slice
+func makeIntSet(ints []int) (s *intsets.Sparse) {
+	s = new(intsets.Sparse)
+	for _, v := range ints {
+		s.Insert(v)
+	}
+	return s
+}
+
+// updateHealStats - update statistics based on given heal result
+func (s *healStatistics) updateHealStats(r healResultItem) {
+	switch {
+	case r.Type == "minioMeta" && r.Detail == "format":
+		s.FormatHealResult = r
+
+	case r.Type == "bucket":
+		s.NumBucketsScanned++
+		avSet := makeIntSet(r.DiskInfo.AvailableOn)
+		for i := 1; i <= s.NumDisks; i++ {
+			if avSet.Has(i) {
+				s.BucketsAvailableByDisk[i]++
+			} else {
+				s.BucketsMissingByDisk[i]++
+			}
+		}
+
+		for _, diskNo := range r.DiskInfo.HealedOn {
+			s.BucketsHealedByDisk[diskNo]++
+		}
+
+	case r.Type == "object" || r.Type == "bucketMeta":
+		s.NumObjectsScanned++
+		avSet := makeIntSet(r.DiskInfo.AvailableOn)
+		for i := 1; i <= s.NumDisks; i++ {
+			if avSet.Has(i) {
+				s.ObjectsAvailableByDisk[i]++
+			} else {
+				s.ObjectsMissingByDisk[i]++
+			}
+		}
+
+		for _, diskNo := range r.DiskInfo.HealedOn {
+			s.ObjectsHealedByDisk[diskNo]++
+		}
+
+		// available part-count
+		avPC := len(r.DiskInfo.AvailableOn)
+		s.ObjectsByAvailablePC[avPC]++
+
+		// healed part-count
+		hdPC := len(r.DiskInfo.HealedOn)
+		s.ObjectsByAvailablePC[hdPC]++
+
+	default:
+		err := fmt.Errorf(
+			"Unknown heal result type (this is a bug): %#v", r)
+		errorIf(err, "Error updating heal statistics")
+	}
+}
+
+type healSequenceStatus struct {
+	Summary       healStatusSummary `json:"Summary"`
+	FailureDetail string            `json:"Detail,omitempty"`
+	HealSettings  healSettings      `json:"Settings"`
+
+	updateLock *sync.Mutex
```

Comment:
`updateLock` is a pointer to avoid copying of the lock when the enclosing structure is passed around/returned (by value).